fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java

```diff
@@ -0,0 +1,675 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.rewrite;
+
+import org.apache.doris.analysis.Analyzer;
+import org.apache.doris.analysis.BinaryPredicate;
+import org.apache.doris.analysis.CompoundPredicate;
+import org.apache.doris.analysis.Expr;
+import org.apache.doris.analysis.InPredicate;
+import org.apache.doris.analysis.IsNullPredicate;
+import org.apache.doris.analysis.JoinOperator;
+import org.apache.doris.analysis.LiteralExpr;
+import org.apache.doris.analysis.SlotRef;
+import org.apache.doris.analysis.TupleId;
+import org.apache.doris.common.AnalysisException;
+import org.apache.doris.common.Pair;
+
+import com.google.common.collect.Sets;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Set;
+import java.util.Map;
+
+/**
+ * The function of this rule is to derive a new predicate based on the current predicate.
+ * eg.
+ * t1.id = t2.id and t2.id = t3.id and t3.id = 100;
+ * -->
+ * t1.id = 100 and t2.id = 100 and t3.id = 100;
+ *
+ * 1. Register a new rule InferFiltersRule and add it to GlobalState.
+ * 2. Traverse Conjunct to construct on/where equivalence connection, numerical connection and isNullPredicate.
+ * 3. Use Warshall to infer all equivalence connections
```

Comment:
It's better to add a link to the "Warshll" algorithm, i.e., a link in Wikipedia, which people could refer to for more details.

Reply:
done